[WIP] Report requires3

kbrock opened the pull request:
Alt to #3391

With references and includes, rails just threw them all into the query.
This was slow, and caused issues when they were polymorphic.
If you included large tables like taggings, then it really blew up

TODO:

- [ ] still prune out taggings, virtual_attributes, and taggings from the references/eager_load tables.

kbrock commented:
update:
- fixed 4 remaining specs

kbrock commented:
update:
- prunes out tags and taggings
- include regular joins in rbac inline view (only used by services)
- changed includes to references for rbac inline view (brings back fewer records)

kbrock commented:
update:
- focused just on models using rbac
- added a few more tests
- added back in the `includes().requires()` broken tests.

I have another PR with 5-10 commits. They were taken out of this and the first rendition of this feature. We can get away with punting on those for now

kbrock commented:
out of band:
TODO:
- ensure there are tests for each of the changed tenant scopes
- verify the scope in #23607 doesn't need to come over here
- Move the `references()` filtering from `Rbac.filterer` to `MiqReport::Generator`

Fryguy commented on lib/rbac/filterer.rb:
```diff
@@ -424,25 +422,45 @@ def select_from_order_columns(columns)
+    def is_collection?(klass, includes)
+      return false unless includes.present?
+
+      Array(includes).each do |association, value|
+        reflection = klass.reflect_on_association(association)
+        return true if reflection&.collection? || !reflection.polymorphic? && is_collection?(reflection.klass, value)
+      end
+      false
```
Simpler?

```suggestion
      Array(includes).any? do |association, value|
        reflection = klass.reflect_on_association(association)
        reflection&.collection? || !reflection.polymorphic? && is_collection?(reflection.klass, value)
      end
```

kbrock replied:
this is a `break` from the iterator.

I will see if we can convert to a `detect?`

kbrock replied:
thanks. this turned out well

Fryguy commented on lib/extensions/ar_references.rb:
```diff
@@ -7,15 +11,45 @@ def includes_to_references(inc)
+            byebug if v_tables.nil?
+            ret[n] = v_tables || {}
+          else
+            byebug unless v.blank?
```
byebug

Fryguy commented on app/models/miq_report/generator.rb:
```diff
@@ -118,16 +103,15 @@ def invent_includes
-    col_order.each_with_object({}) do |col, ret|
+    (Array(col_order) + Array(sortby)).each_with_object({}) do |col, ret|
       next unless col.include?(".")
+      next if col =~ /virtual_custom/
```
```suggestion
      next if col.include?("virtual_custom")
```

kbrock commented:
update:
- dropped `includes_to_references`
- `prune_references` now just returns a hash
- `add_joins` changed back to silently eats polymorphics (had changed to throw exception but rolled back)
- specs around prune_references
- added comment 

update:
- converted regular expression to `include?()`

Fryguy commented:
@miq-bot cross-repo-test /all

Fryguy commented:
Backported to `spassky` in commit 6cc541592815fc500066bbd449b5d99a7e64deff.

```text
commit 6cc541592815fc500066bbd449b5d99a7e64deff
Author: Jason Frey <fryguy9@gmail.com>
Date:   Tue Oct 7 10:35:48 2025 -0400

    Merge pull request #23601 from kbrock/report_requires3
    
    convert MiqReport from references to eager_load
    
    (cherry picked from commit 7cddb9a3f4bde9885f03db5238d69f20375b7499)
```
